monai/metrics/seg_metric_utils.py

```diff
@@ -0,0 +1,220 @@
+        bin_mode: [``None``, ``"threshold"``, ``"mutually_exclusive"``]
+            Binarization method, if specified, a binarization manipulation will be employed for `y_pred` before
+            computation. Defaults to None.
+
+            - ``"threshold"``, a single threshold or a sequence of thresholds should be set.
+            - ``"mutually_exclusive"``, `y_pred` will be converted by a combination of `argmax` and `to_onehot`.
+        bin_threshold: the threshold for binarization, can be a single value or a sequence of
+            values that each one of the value represents a threshold for a class.
+        include_background: whether to include computation on the first channel of the predicted output.
+            Defaults to True.
+
+    Raises:
+        AssertionError: When ``y_pred`` only has one dimebsion.
+        AssertionError: When ``y`` and ``y_pred`` have different shapes before output.
+
+    """
+
+    # convert y into one-hot version if needed
+    assert y_pred.ndimension() > 1, "y_pred should at least has two dimensions batch and n_classes."
+    num_classes = y_pred.shape[1]
+    if to_onehot_y:
+        y = do_onehot(y, num_classes)
+    # do activation for y_pred if needed
+    if activation is not None:
+        y_pred = do_activation(y_pred, activation=activation)
+    # binarization for y_pred if needed
+    if bin_mode is not None:
+        y_pred = do_binarization(y_pred, bin_mode=bin_mode, bin_threshold=bin_threshold)
+    # remove background if needed
+    if not include_background:
+        y = y[:, 1:] if y.shape[1] > 1 else y
+        y_pred = y_pred[:, 1:] if y_pred.shape[1] > 1 else y_pred
+    assert y.shape == y_pred.shape, "data shapes of y_pred and y do not match."
+    y = y.float()
+    y_pred = y_pred.float()
+    return y_pred, y
+
+
+def do_metric_reduction(
+    f: torch.Tensor,
+    reduction: Union[MetricReduction, str] = MetricReduction.MEAN,
+):
+    """
+    This function is to do the metric reduction for calculated metrics of each example's each class.
+    Args:
+        f: a tensor that contains the calculated metric scores per batch and
+            per class. The first two dims should be batch and class.
+        reduction: {``"none"``, ``"mean"``, ``"sum"``, ``"mean_batch"``, ``"sum_batch"``,
+        ``"mean_channel"``, ``"sum_channel"``}
+        Define the mode to reduce computation result of 1 batch data. Defaults to ``"mean"``.
+
+    Raises:
+        ValueError: When ``reduction`` is not one of
+            ["mean", "sum", "mean_batch", "sum_batch", "mean_channel", "sum_channel" "none"].
+    """
+
+    # some elements might be Nan (if ground truth y was missing (zeros))
+    # we need to account for it
+    nans = torch.isnan(f)
+    not_nans = (~nans).float()
+    f[nans] = 0
+
+    t_zero = torch.zeros(1, device=f.device, dtype=torch.float)
+    reduction = MetricReduction(reduction)
+
+    if reduction == MetricReduction.MEAN:
```

Comment:
Could you please help also test the reduction ops for the classification task?
Thanks.

Reply:
I see, I just removed the preprocess part and update the code. My next step is to add classification support. I will switch to a formal pr after finishing all places.